github-backup.py

```diff
@@ -27,118 +42,439 @@ def main():
+    parser.add_argument('--gists',
+                        action='store_true',
+                        dest='include_gists',
+                        help='include gists in backup [*]')
+    parser.add_argument('--starred-gists',
+                        action='store_true',
+                        dest='include_starred_gists',
+                        help='include starred gists in backup [*]')
+    parser.add_argument('--releases',
+                        action='store_true',
+                        dest='include_releases',
+                        help='include release information, not including assets or binaries'
+                        )
+    parser.add_argument('--assets',
+                        action='store_true',
+                        dest='include_assets',
+                        help='include assets alongside release information; only applies if including releases')
 
     return parser
 
-
-def process_repo(repo, args):
-    LOGGER.info("Processing repo: %s", repo.full_name)
-
-    dir = args.backupdir + '/' + args.prefix + repo.name + args.suffix
-    config = "%s/%s" % (dir, "config" if args.mirror else ".git/config")
-
-    if not os.access(config, os.F_OK):
-        LOGGER.info("Repo doesn't exists, lets clone it")
-        clone_repo(repo, dir, args)
-    else:
-        LOGGER.info("Repo already exists, let's try to update it instead")
-        update_repo(repo, dir, args)
-
-
-def clone_repo(repo, dir, args):
-    if args.type == 'http':
-        url = repo.clone_url
-    elif args.type == 'ssh':
-        url = repo.ssh_url
-    elif args.type == 'git':
-        url = repo.git_url
-
-    git_args = [url, os.path.basename(dir)]
-    if args.mirror:
-        git_args.insert(0, '--mirror')
-
-    git("clone", git_args, args.git, args.backupdir)
-
-
-def update_repo(repo, dir, args):
-    # GitHub => Local
-    # TODO: use subprocess package and fork git into
-    #       background (major performance boost expected)
-    if args.mirror:
-        git("fetch", ["--prune"], args.git, dir)
-    else:
-        git("pull", gargs=args.git, gdir=dir)
-
-    # Fetch description and owner (useful for gitweb, cgit etc.)
-    if repo.description:
-        git("config", ["--local", "gitweb.description",
-            repo.description.encode("utf-8")], gdir=dir)
-
-    if repo.owner.name and repo.owner.email:
-        owner = "%s <%s>" % (repo.owner.name.encode("utf-8"),
-                             repo.owner.email.encode("utf-8"))
-        git("config", ["--local", "gitweb.owner", owner], gdir=dir)
-
-    git("config", ["--local", "cgit.name", str(repo.name)], gdir=dir)
-    git("config", ["--local", "cgit.defbranch", str(repo.default_branch)], gdir=dir)
-    git("config", ["--local", "cgit.clone-url", str(repo.clone_url)], gdir=dir)
+def fetch_url(url, outfile):
+    headers = {
+        "User-Agent": "PyGithub/Python"
+    }
+    with open(outfile, 'w') as f:
+        f.write(requests.get(url, headers=headers).content)
```

Comment:
I am getting an exception here:

```
INFO:github-backup:Cannot backup keys with unauthenticated account, ignoring...
INFO:github-backup:Processing account: stv0g
INFO:github-backup:    Getting starred repository list
Traceback (most recent call last):
  File "./github-backup.py", line 513, in <module>
    main()
  File "./github-backup.py", line 132, in main
    process_account(gh, account, args)
  File "./github-backup.py", line 269, in process_account
    fetch_url(account.starred_url, os.path.join(dir, 'starred.json'))
  File "./github-backup.py", line 249, in fetch_url
    f.write(requests.get(url, headers=headers).content)
TypeError: write() argument must be str, not bytes
```

Reply:
Probably due to Python2/3 compatability...

Reply:
Yep, but its fixed now for both Python 2/3.